app/controllers/link_controller.rb

```diff
@@ -0,0 +1,63 @@
+require 'net/https'
+require 'uri'
+require 'json'
+require_relative '../../models/link'
+require_relative '../../app/controllers/application_controller'
+require_relative '../../models/category'
+require_relative '../../services/bing_service'
+require 'warden'
+
+class LinkController < ApplicationController
+  bing_search_api = BingService::BingWebSearchApi.new
+
+  get '/category_links/:category_id' do
+    redirect '/login' unless current_user
+    @current_user = current_user
+
+    term = @current_user.categories.find_by(id: params[:category_id].to_i)
+    parsed_json = bing_search_api.links(term)
+
+    @category_links = parsed_json['webPages']['value']
+
+    haml :'category/category_links'
+  end
+
+  get '/category_link/:category_id' do
+    redirect '/login' unless current_user
+    @current_user = current_user
+
+    term = Category.where(id: params[:category_id].to_i).first
+    parsed_json = bing_search_api.links(term)
+
+    result_set = parsed_json['webPages']['value']
+    # result_set.shuffle!
+
+    category = @current_user.categories.find_by(id: params[:category_id].to_i)
+
+    link_name = category.links.find_by(link_name: result_set[0]['displayUrl'])
+
+    until link_name.nil?
+      result_set.shuffle!
+      link_name = category.links.find_by(link_name: result_set[0]['displayUrl'])
```

Comment:
Bug: **Possible Infinite Loop.** The reason you cannot see the bug can be attributed to the error I've highlighted above.

Imagine a scenario where `result_set[0]['displayUrl']` has already been visited, and thus saved in your database. In such a situation, you will have an infinite loop. Because you are providing the exact same search result (the one indexed at **0**). I'd have expected you to increment that index on every iteration.

Please review this logic.